# 😎 오류 처리 - 독서 정리

깨끗한 코드와 오류 처리는 확실히 연관성이 있다. 상당수 코드 기반은 전적으로 오류 처리 코드에 좌우된다.

- 흩어진 오류 처리 코드 때문에, 실제 코드가 하는 일을 파악하기가 거의 불가능하다는 의미
- 오류 처리는 중요하다.
- **오류 처리 코드로 인해, 프로그램 논리를 이해하기 어려워진다면 깨끗한 코드라 부르기 어렵다.**



## 😱 오류 코드보다 예외를 사용하자

- 예외란? ***프로그램 실행 중에 정상적인 프로그램의 흐름에 어긋나는 이벤트***

- 함수를 호출한 즉시 오류를 확인하는 코드는 구리다. 차라리 오류가 발생하면 예외를 던지는 게 낫다. 논리가 오류 처리 코드와 뒤섞이지 않으니까

  ```java
  // BAD (오류 확인 코드)
  public class DeviceController{
    ...
    public void sendShutDown() {
      DeviceHandle handle = getHandle(DEV1); 
      // 디바이스 상태 점검
      if (handle != DeviceHandle.INVALID) {
        // 레코드 필드에 디바이스 상태 저장
        retrieveDeviceRecord(handle);
        // 디바이스가 일시정지 상태가 아니라면 종료
        if (record.getStatus() != DEVICE_USPENDED) {
          pauseDevice(handle);
          clearDeviceWorkQueue(handle);
          closeDevice(handle);
        } else {
          logger.log("Device suspended. Unable to shut down");
        }
      } else {
        logger.log("Invalid handle for: " + DEV1.toString());
      }
    }
    ...
  }
  
  // GOOD (예외 사용)
  // 디바이스 종료 알고리즘과 오류 처리 알고리즘이 분리됐다.
  public class DeviceController{
    ...
    public void sendShutDown() {
      try {
        tryToShutDown();
      } catch (DeviceShutDownError e) {
        logger.log(e);
      }
    }
    
    private void tryToShutDown() throws DeviceShutDownError {
      DeviceHandle handle = getHandle(DEV1);
      DeviceRecord record = retrieveDeviceRecord(handle);
      
      pauseDevice(handle);
      clearDeviceWorkQueue(handle);
      closeDevice(handle);
    }
    
    private DeviceHandle getHandle(DeviceID id) {
      throw new DeviceShutDownError("Invalid handle for: " + id.toString());
      ...
    }
    ...
  }
  ```

   



## 😱 Try-Catch-Finally 문부터 작성하라

 예외가 발생할 코드를 짤 때는 try-catch-finally 문으로 시작하는 편이 낫다.

- Try 블록에서 무슨 일이 생기든지, 호출자가 기대하는 상태를 정의하기 쉬워진다.
- TDD에서 먼저 강제로 예외를 일으키는 테스트 케이스를 작성한 후 테스트를 통과하게 코드를 작성하면, 자연스럽게 try 블록의 트랜잭션 범위부터 구현하게 되어 범위 내에서 트랜잭션 본질을 유지하기 쉬워진다.



## 😱 미확인(unchecked) 예외를 사용하라

예전에 메서드를 선언할 때는, 메서드가 반환할 예외를 모두 열거했다.

이러한 **확인된 오류** 가, 치르는 비용에 상응하는 이익을 제공하는지 철저히 따져야 한다.

- 확인된 예외는  OCP(개방, 폐쇄의 원칙)을 위반한다.
  - 하위 단계에서 코드를 변경하면 상위 단계 메서드 선언부를 전부 고쳐야 한다.
- **때로는 확인된 예외도 유용하지만, 일반적 애플리케이션은 의존성이라는 비용이 이익보다 크다.**



#### [ 확인된 예외(checked exception) ]

- 컴파일 단계에서 확인되며 **반드시 별도로 처리**해야 하는 예외
  - 프로그램 내에서 통제할 수 없는 조건 때문에 발생 (Ex) 예를들어 존재하지 않는 외부 파일을 사용할 때 발생)
    - IOException
    - SQLException
- 파일 열기와 같이 정확한 코드로 구현했음에도, 외부 환경(파일이 없는 상황 등)에 따라 발생 가능
- 예외처리를 구현하지 않으면 컴파일 에러 발생 (컴파일 시 확인해서 확인된 예외)
- RuntimeException 이외의 예외들

#### [ 미확인 예외(unchecked exception) ]

- 실행 단계에서 확인되며 명시적인 처리를 강제하지는 않는 예외 (컴파일 단계에서 확인하지 않음 = 미확인 예외)
  -  RuntimeExcepion과 직간접적으로 관련된 객체
  - 잘못 구현된 프로그램 코드 때문에 발생
  - 적절한 예외처리를 안할 경우, 프로그램 강제 종료를 일으킴
    - NullPointerExceiption
    - IllegalArgumentException
    - IndexOutOfBoundException
    - SystemException



## 😱 예외에 의미를 제공하라

오류가 발생한 원인과 위치를 찾기 쉽도록, 예외를 던질 때는 전후 상황을 충분히 덧붙인다.

- 오류 메시지에 정보를 담아 예외와 함께 던진다.
- 실패한 연산 이름과 실패 유형도 언급
- 애플리케이션이 로깅 기능을 사용한다면, catch 블록에서 오류를 기록하도록 충분한 정보를 넘겨준다.



## 😱 호출자를 고려해 예외 클래스를 정의하라

오류의 분류 방법은 수없이 많다. 

- 컴포넌트 분류
- 유형으로 분류
  - 디바이스 실패
  - 네트워크 실패
  - 프로그래밍 오류

등.

하지만 애플리케이션에서 오류 정의시 가장 중요한 관심사는, **오류는 잡아내는 방법** 이 되어야 한다.

아래 코드는 코드를 간결하게 고치기가 쉽다. 호출하는 라이브러리 API를 감싸면서 예외 유형 하나를 반환하면 된다.

```java
LocalPort port = new LocalPort(12);
try {
  port.open();
} catch (PortDeviceFailure e) {
  reportError(e);
  logger.log(e.getMessage(),e);
} finally {
  ...
}
```

여기서 LocalPort 클래스는 단순히 ACMEPort 클래스가 던지는 예외를 잡아 변환하는 **감싸기(wrapper) 클래스**일 뿐이다.

```typescript
public class LocalPort {
	private ACMEPort innerPort;
  
  public LocalPort(int portNumber) {
    innerPort = new ACMEPort(portNumber);
  }
  
  public void open() {
    try {
      innerPort.open();
    } catch (DeviceResponseException e) {
      throw new PortDeviceFailure(e);
    } catch (ATM1212UnlockedException e) {
      throw new PortDeviceFailure(e);
    } catch (GMXError e) {
      throw new PortDeviceFailure(e);
    }
  }
  
  ...
}
```

#### 감싸기(wrapper) 기법

외부 API를 사용할 때 좋다. 

- 외부 라이브러리와 프로그램 사이의 의존성이 크게 줄어듦
- 테스트가 쉽다
- 특정 업체의 API 설계 방식에 발목 잡히지 않는다.
  - 네트워크에 적용해볼까 (에러코드 상관 없이,,)



## 😱 정상 흐름을 정의하라

위와 같은 방식으로, 대부분 오류 처리가 잘 분리된 코드가 나오지만, 때로는 중단이 적합하지 않은 때도 있다.

-> 특수 사례 패턴(클래스를 만들거나 객체를 조작해 특수 사례를 처리하는 방식)으로 처리



## 😱 null을 반환하지 마라

null을 반환하는 습관을 버리자. 한줄 건거 하나씩 null을 확인하는 코드로 가득한 애플레이션은, 일거리를 늘릴 뿐만 아니라 호출자에게 문제를 떠넘긴다.

- 메서드에서 null을 반환하고픈 유혹이 든다면, 그 대신 예외를 던지거나 특수사례객체를 반환한다.
- 사용하려는 외부 API가 null을 반환하면, wrapper 메서드를 구현해 예외를 던지거나 특수사례객체를 반환한다.



```java
// bad case
// getEmployees는 null도 반환한다.
List<Employee> employees = getEmployees();
if (employees != null) {
  for(Employee e : employees) {
    totalPay += e.getPay();
  }
}

// good case
// getEmployees를 변경해 빈 리스트를 반환한다면 코드가 훨씬 깔끔해짐
// NullPointerException이 발생할 가능성도 줄어든다.
List<Employee> employees = getEmployees();
  for(Employee e : employees) {
    totalPay += e.getPay();
  }
}

public List<Employee> getEmployees() {
  if (..직원이 없다면 ..)
    return Collections.emptyList();
}
```





## 😱 null을 전달하지 마라

메서드에서 null을 반환하는 것도 나쁘지만, **메서드로 null을 전달하는 방식은 더 나쁘다.**

- 메서드로 null을 전달하는 코드는 최대한 피한다. (정상적 인수로 null을 기대하는 API가 아니라면)
  - 실행 오류가 발생할 가능성이 있다.
  - 대다수 프로그래밍 언어는 호출자가 실수로 넘기는 null을 적절히 처리하지 못한다.





# 😎 오류 처리 - 강의 복습

## 01. 예외 처리 방식

오류 코드를 리턴하지 말고, 예외를 던져라

- 옛날에는 오류를 나타낼 때 에러코드를 던졌다.
- 하지만 예외를 던지는 것이 명확하고, 처리 흐름이 깔끔해진다.

```java
// BAD
public class DeviceController {
  ...
  public void sendShutDown(){
    if (handle != DeviceHandle.INVALID){ // here
      ...
      if (record.getStatus() != DEVICE_SUSPENDED) { // here
        ...
      }
    } 
  }
}

// GOOD (정말 굿은 아님)
// (1) 오류가 발생한 부분에서 예외를 던진다. (별도로 개발자의 처리가 필요한 예외이면 checked exception으로 던진다.)
// (2) checked exception에 대한 예외처리를 하지 않는다면, 상위에서 처리할 수 있도록 메서드 선언부에 throws를 명시해야 한다.
// (3) 예외를 처리할 수 있는 곳에서 catch 하여 처리한다.
public class DeviceController {
  ...
  public void sendShutDown(){
    // here (3)
    try {
      tryToShutDown();
    } catch( DeviceShutDownError e) {
      logger.log(e);
    }
  }
  
  private void tryToShutDown() throws DeviceShutDownError { // here (2)
    DeviceHandle handle = getHandle(DEV1);
    ...
  }
  
  private DeviceHandle getHandle(DeviceID id) {
    ...
    // Device를 정상적으로 가져올 수 없다면
    throw new DeviceShutDownError("Invalid handler for: " + id.toString()); // here(1)
    ...
  }
}
```



## 02. Unckecked Exception을 활용하라

### Checked vs Unchecked Exception

![](https://user-images.githubusercontent.com/45676906/105691109-2cda9400-5f40-11eb-9003-a14873c2eaf2.png)

- **Exception**을 바로 상속하면 Checked Exception
  - 명시적인 예외처리가 필요하다.
  - Ex) IOException, SQLException
- **RuntimeException**을 상속하면 UnckeckedException
  - 명시적인 예외처리가 필요하지 않다.
    - 런타임(실행) 중에 발생했기 때문에, 정상적으로 실행할 수 없다는 에러이기 때문에, 이 때는 실행중에 별도로 발생한 에러라 처리를 할 수 없어서 명시적인 처리가 필요하지 않음.
  - Ex) NullPointerException, IllegalArgumentException, IndexOutOfBoundException

> 사용자가 직접 구현하는 Unchecked throwable은 모두 RuntimeException의 하위 클래스여야 한다.
>
> Exception, RuntimeException, Error를 상속하지 않는 throwable을 만들 수도 있지만, 이러한 throwable은 정상적인 사항보다 나을 게 하나도 없으면서 API 사용자를 헷갈리게 할 뿐이므로 절대로 사용하지 말자.

### Checked Exception이 나쁜 이유

```java
// (1) 특정 메소드에서 checked exception을 throw하고 상위 메소드에서 그 exception을 catch한다면,
//     모든 중간단계 메소드에 exception을 throws 해야 한다.
// (2) OCP(개방 폐쇄 원칙) 위배
//     상위 레벨 메소드에서 하위 레벨 메소드의 디테일에 대해 알아야 하기 때문에 OCP 원칙에 위배된다.
// (3) 필요한 경우 checked exception을 사용해야 되지만 일반적인 경우 득보다 실이 많다.
public class DeviceController {
  ...
  public void sendShutDown(){
    // here (3)
    try {
      tryToShutDown();
    } catch( DeviceShutDownError e) {
      logger.log(e);
    }
  }
  
  private void tryToShutDown() throws DeviceShutDownError { // here (2)
    DeviceHandle handle = getHandle(DEV1);
    ...
  }
  
  private DeviceHandle getHandle(DeviceID id) {
    ...
    // Device를 정상적으로 가져올 수 없다면
    throw new DeviceShutDownError("Invalid handler for: " + id.toString()); // here(1)
    ...
  }
}
```

### Unchecked Exception을 사용하자

> 확인된 예외를 지원하지 않음에도 불구하고, C#, C++, 파이썬, 루비는 안정적인 소프트웨어를 구현하기에 무리가 없다.



## 03. Exception 잘 쓰기

### - 예외에 메시지 담기

예외에 의미있는 정보 담기

```java
throw new DeviceShutDownError("Invalid handle for: " + id.toString());
```

- 오류가 발생한 원인과 위치를 찾기 쉽도록, 예외를 던질 때는 전후 상황을 충분히 덧붙인다.
- 실패한 연산 이름과 유형 등 정보를 담아 예외를 던진다.

### - exception wrapper

```java
// BAD
// 수많은 checked exception이 발생
// checked exception은 반드시 처리를 해줘야 하지만, 
// 그러나 로그를 찍을 뿐 할 수 있는 일이 없다.
ACMEPort port = new ACMEPort(12);
try {
  port.open();
} catch (DeviceResponseException e) {
  reportPortError(e);
  logger.log("Device response exception",e);
} catch (ATM1212UnlockedException e) {
  reportPortError(e);
  logger.log("Unlock exception",e);
} catch (GMXError e) {
  reportPortError(e);
  logger.log("Device response exception");
} finally {
  ...
}

// Good
// 예외를 감싸는 wrapper 클래스를 만든다.
LocalPort port = new LocalPort(12);
try {
  port.open();
} catch(PortDeviceFailure e){
  reportError(e);
  logger.log(e.getMessage(), e);
} finally {
  ...
}

public class LocalPort {
  private ACMEPort innerPort;
  public LocalPort(int portNumber) {
    innerPort = new ACMEPort(portNumber);
  }
  
  // port.open()시 발생하는 checked exception들을 감싸도록 port를 가지는 LocalPort 클래스를 만든다.
  // port.open()이 던지는 checked exceiption들을 하나의 PortDeviceFailure exception으로 감싸서 던진다.
  public void open() {
    try {
      innerPort.open();
    } catch (DeviceResponseException e) {
			  throw new PortDeviceFailure(e);
		} catch (ATM1212UnlockedException e) {
			  throw new PortDeviceFailure(e);
		} catch (GMXError e) {
			  throw new PortDeviceFailure(e);
  	}
  }
}
```



## 04. 실무 예외 처리 패턴

**getOrElse**로 안되면 **getOrElseThrow** 로 처리한다.

- getOrElse
  - 예외 대신 기본 값을 리턴한다.
    1. null이 아닌 기본 값
    2. 그게 아니라면, 도메인에 맞는 기본 값
- getOrElseThrow
  - null 대신 예외를 던진다. (기본값이 없다면)



```java
// 👹 case1. getOrElse - null이 아닌 기본값을 리턴한다.

// BAD
// null을 리턴하면 이후 모든 코드에서 null 체크가 있어야 하는데, 다른 방법이 없을까?
List<Employee> employees = getEmployees();
if (employees != null){
  for(Employee e : employees) {
    totalPay += e.getPay();
  }
}

// GOOD
// 복수형 데이터를 가져올 때는 데이터의 없음을 의미하는 컬렉션을 리턴하면 된다,
// null 보다 size가 0인 컬렉션이 훨씬 안전하다.
List<Employee> employees = getEmployees();
for(Employee e : employees) {
  totalPay += e.getPay();
}

public List<Employee> getEmployees() {
  if ( .. there are no employees .. ) {
    return Collections.empltyList();
  }
}



// 빈 컬렉션, 빈 문자열을 적용할 수 없는 경우라면?
// 👹 case2. getOrElse - 도메인에 맞는 기본값을 가져온다.

// BAD
// 호출부에서 예외처리를 통해 userLevel값을 처리하는데, 이렇게 되면 코드를 계속 읽어나가면서 논리적 흐름이 끊긴다.
UserLevel userLevel = null;
try {
  User user = userRepository.findByUserId(userId); // userNotFoundException
  userLevel = user.getUserLevel();
} catch (UserNotFoundException e) {
  userLevel = UserLevel.BASIC;
}

// GOOD
// 예외 처리를 데이터를 제공하는 쪽에서 처리해 호출부 코드가 심플해진다.
// 코드를 읽어가며 논리적인 흐름이 끊기지 않는다.
// 도메인에 맞는 기본값을 도메인 서비스에서 관리한다.

// 호출부
UserLevel userLevel = userService.getUserLevelOrDefault(userId);

// 제공부
public class UserService {
  private static final UserLevel USER_BASIC_LEVEL = UserLevel.BASIC;
  
  public UserLevel getUserLevelOrDefaul(Long userId) {
    try {
      User user = userRepository.findByUserId(userId);
      return user.getUserLevel();
    } catch (UserNotFoundException e) {
      return USER_BASIC_LEVEL;
    }
  }
}



// 도메인에 맞는 기본값조차 없다면?
// 👹 case3. getOrElseThrow - null 대신 예외를 던진다
// null 체크 지옥에서 벗어나기 위해

// BAD
// null 체크 빠진 부분이 발생해도 알아차리기 힘들 수 있는 등, 코드 가독성이 현저히 떨어진다.
public void registerItem(Item item) {
  if(item != null) {
    ItemRegistry registry = peristenStore.getIemRegistry();
    if (registry != null) {
      Item existing = registry.getItem(item.getID());
      if (existing.getBillingPeriod().hasRetailOwner()) {
        existing.register(item);
      }
    }
  }
}

// BAD
// 호출부에서 매번 null 체크해야 하는데, 가독성뿐 아니라 안정성도 떨어진다.
User user = userRepository.findByUserId(userId);
if (user != null){
  // user를 이용한 처리
}

// GOOD
// 호출부 가독성이 올라간다.
// 데이터가 없는 경우엔 예외를 던진다.
// 매번 null 체크할 필요가 없어 안전하게 데이터 처리할 수 있음.

// 호출부
User user = userService.getUserOrElseThrow(userId);

// 제공부
public class UserService {
  private static final UserLevel USER_BASIC_LEVEL = UserLevel.BASIC;
  
  public User getUserOrElseThrow(Long userId) {
    User user = userRepository.findByUserId(userId);
    if(user == null) {
      throw new IllegalArgumentException("User is not found. userId = " + userId)
    }
    return user;
  }
}

```

### 실무에서는 보통 자신의 예외를 정의한다.

- 에러 로그에서 stacktrace 해봤을 때 우리가 발생시킨 예외라는 것을 바로 인지할 수 있다
- 다른 라이브러리에서 발생한 에러와 섞이지 않는다. 우리도 IllegalArgumentException을 던지는 거보다 우리 예외로 던지는게 어느 부분에서 에러가 났는지 파악하기에 용이하다.
- 우리 시스템에서 발생한 에러의 종류를 나열할 수 있다.

```java
public class MyProjectException extends RuntimeException {
  private MyErrorCode errorCode;
  private String errorMessage;
  
  public MyProjectException(MyErrorCode errorCode) {
    // ..
  }
  
  public MyProjectExceiption(MyErrorCode errorCode, String errorMessage) {
    // ..
  }
}

public enum MyErrorCode {
  private String defaultErrorMessage;
  
  INVALID_REQUEST("잘못된 요청입니다."),
  DUPLLICATED_REQUEST("기존 요청과 중복되어 처리할 수 없습니다."),
  // ,,,
  INTERNAL_SERVER_ERROR("처리 중 에러가 발생했습니다.");
}

// 호출부
if(request.getUserName() == null) {
  throw new MyProjectException(ErrorCode.INVALID_REQUEST, "userName is null");
}
```



## 05. 오픈소스 속 Exception 살펴보기

